[Merged by Bors] - Add cli option for voluntary exits

pawanjay176 opened the pull request:
## Issue Addressed

Resolve #1652 

## Proposed Changes

Adds a cli option for voluntary exits. The flow is similar to prysm's where after entering the password for the validator keystore (or load password from `secrets` if present) the user is given multiple warnings about the operation being irreversible, then redirected to the docs webpage(not added yet) which explains what a voluntary exit is and the consequences of exiting and then prompted to enter a phrase from the docs webpage as a final confirmation. 

Example usage
```
$ lighthouse --testnet zinken account validator exit --validator <validator-pubkey> --beacon-node http://localhost:5052

Running account manager for zinken testnet                                                                                                          
validator-dir path: "..."

Enter the keystore password:  for validator in ...

Password is correct

Publishing a voluntary exit for validator: ...              
WARNING: This is an irreversible operation                                                                                                                    
WARNING: Withdrawing staked eth will not be possible until Eth1/Eth2 merge Please visit [website] to make sure you understand the implications of a voluntary exit.            
                                                                                                                                             
Enter the phrase from the above URL to confirm the voluntary exit:
Exit my validator
Published voluntary exit for validator ...
```

## Additional info

Not sure if we should have batch exits (`--validator all`) option for exiting all the validators in the `validators` directory. I'm slightly leaning towards having only single exits but don't have a strong preference.

pawanjay176 commented:
Tested this out on a few of my medalla validators and it works as expected.

paulhauner commented on book/src/voluntary-exit.md:
```diff
@@ -0,0 +1,72 @@
+# Voluntary exits
+
+A validator may chose to voluntary stop performing duties (proposing blocks and attesting to blocks) by submitting
```
```suggestion
A validator may chose to voluntarily stop performing duties (proposing blocks and attesting to blocks) by submitting
```

paulhauner commented on book/src/voluntary-exit.md:
```diff
@@ -0,0 +1,72 @@
+A validator may chose to voluntary stop performing duties (proposing blocks and attesting to blocks) by submitting
+a voluntary exit transaction to the beacon chain.
+
+A validator can initiate a voluntary exit provided that the validator is currently active (has not been slashed) and has been active for at least 256 epochs (~27 hours) since it has been activated.
```
```suggestion
A validator can initiate a voluntary exit provided that the validator is currently active, has not been slashed and has been active for at least 256 epochs (~27 hours) since it has been activated.
```

paulhauner commented on book/src/voluntary-exit.md:
```diff
@@ -0,0 +1,72 @@
+
+> Note: After initiating a voluntary exit, the validator will have to keep performing duties until it has successfully exited to avoid penalties.
+
+It takes at minimum 5 epochs (32 minutes) for a validator to exit after initiating a voluntary exit.
```
```suggestion
It takes at a minimum 5 epochs (32 minutes) for a validator to exit after initiating a voluntary exit.
```

paulhauner commented on book/src/voluntary-exit.md:
```diff
@@ -0,0 +1,72 @@
+> Note: After initiating a voluntary exit, the validator will have to keep performing duties until it has successfully exited to avoid penalties.
+
+It takes at minimum 5 epochs (32 minutes) for a validator to exit after initiating a voluntary exit.
+This number can be much higher depending on how many other validators are in queue to exit.
```
```suggestion
This number can be much higher depending on how many other validators are queued to exit.
```

paulhauner commented on book/src/voluntary-exit.md:
```diff
@@ -0,0 +1,72 @@
+
+## Withdrawal of exited funds
+
+Even though users can perform a voluntary exit in phase 0, they **cannot withdraw their exited funds until phase 2**.
```
```suggestion
Even though users can perform a voluntary exit in phase 0, they **cannot withdraw their exited funds at this point in time**.
```

paulhauner replied:
To avoid committing to a withdrawal-is-possible date.

paulhauner commented on book/src/voluntary-exit.md:
```diff
@@ -0,0 +1,72 @@
+
+## Initiating a voluntary exit
+
+In order to initiate a exit, users can use the `lighthouse account validator exit` command.
```
```suggestion
In order to initiate an exit, users can use the `lighthouse account validator exit` command.
```

paulhauner commented on book/src/voluntary-exit.md:
```diff
@@ -0,0 +1,72 @@
+
+The `--validator` flag is used to specify the validator for which the user wants to initiate an exit transaction.
+
+The `--beacon-node` flag is used to specify a beacon chain http endpoint that confirms to the [standard api](https://ethereum.github.io/eth2.0-APIs/) specifications which will be used to validate and propagate the voluntary exit. The default value for this flag is `http://localhost:5052`.
```
```suggestion
The `--beacon-node` flag is used to specify a beacon chain HTTP endpoint that confirms to the [Eth2.0 Standard API](https://ethereum.github.io/eth2.0-APIs/) specifications. That beacon node will be used to validate and propagate the voluntary exit. The default value for this flag is `http://localhost:5052`.
```

paulhauner commented on book/src/voluntary-exit.md:
```diff
@@ -0,0 +1,72 @@
+
+The `--testnet` flag is used to specify a particular testnet (default is `medalla`).
+
+The `exit` command will search for the provided validator in the `validators-dir` (default `~/.lighthouse/{testnet}/validators`) directory and validator password file in `secrets-dir` (default `  ~/.lighthouse/{testnet}/secrets`).
```
```suggestion
The `exit` command will search for the provided validator in the `validators-dir` (default `~/.lighthouse/{testnet}/validators`) directory and for the validator's password file in `secrets-dir` (default `  ~/.lighthouse/{testnet}/secrets`).
```

paulhauner commented on book/src/voluntary-exit.md:
```diff
@@ -0,0 +1,72 @@
+
+The `exit` command will search for the provided validator in the `validators-dir` (default `~/.lighthouse/{testnet}/validators`) directory and validator password file in `secrets-dir` (default `  ~/.lighthouse/{testnet}/secrets`).
+
+If the corresponding  password file is not found in the `secrets-dir`, the user will be prompted for a password on the terminal.
```
```suggestion
If the corresponding password file is not found in the `secrets-dir`, the user will be prompted for a password on the terminal.
```

paulhauner commented on book/src/voluntary-exit.md:
```diff
@@ -0,0 +1,72 @@
+
+If the corresponding  password file is not found in the `secrets-dir`, the user will be prompted for a password on the terminal.
+
+[//]: # (Word this better or remove datadir part entirely)
```
TODO

paulhauner commented on book/src/voluntary-exit.md:
```diff
@@ -0,0 +1,72 @@
+
+In order to initiate a exit, users can use the `lighthouse account validator exit` command.
+
+The `--validator` flag is used to specify the validator for which the user wants to initiate an exit transaction.
```
If we don't just use an EIP-2335 keystore path instead, I think we should specify how the value of this flag identifies a validator.

If we *do* just use an EIP-2335 keystore path, you can ignore this comment.

paulhauner commented on book/src/SUMMARY.md:
```diff
@@ -15,6 +15,7 @@
 * [Validator Management](./validator-management.md)
 	* [Importing from the Eth2 Launchpad](./validator-import-launchpad.md)
     * [Slashing Protection](./slashing-protection.md)
+    * [Voluntary exits](./voluntary-exit.md)
```
```suggestion
    * [Voluntary Exits](./voluntary-exit.md)
```

paulhauner commented on account_manager/src/validator/exit.rs:
```diff
@@ -0,0 +1,440 @@
+pub const CONFIRMATION_PHRASE: &str = "Exit my validator";
+pub const WEBSITE_URL: &str = "https://lighthouse-book.sigmaprime.io/voluntary-exit.html";
+pub const PROMPT: &str =
+    "WARNING: WITHDRAWING STAKED ETH WILL NOT BE POSSIBLE UNTIL ETH1/ETH2 MERGE";
```
```suggestion
    "WARNING: WITHDRAWING STAKED ETH IS NOT PRESENTLY POSSIBLE";
```

Just to avoid committing to a time when withdraws will be permitted, since it's out of our control.

paulhauner commented on account_manager/src/validator/exit.rs:
```diff
@@ -0,0 +1,440 @@
+        validator_index,
+    };
+
+    let signed_voluntary_exit =
```
Just to be super cautious, I would suggest creating this *after* the confirmation.

paulhauner commented on account_manager/src/validator/exit.rs:
```diff
@@ -0,0 +1,440 @@
+    let validator_data = client
+        .get_beacon_states_validator_id(
+            // TODO: verify this is the state that we want to query.
+            StateId::Slot(epoch.start_slot(E::slots_per_epoch())),
```
I think it would be best to use `StateId::Head` here, it's much faster for the BN API to compute.

It should be equivalent, but in some rare circumstances we might find that the validator is not present or active in the head state, which will just throw an error. I find this to be safe.

paulhauner commented on account_manager/src/validator/exit.rs:
```diff
@@ -0,0 +1,440 @@
+    match validator_data.status {
+        ValidatorStatus::Active => {
+            let eligible_epoch =
+                validator_data.validator.activation_epoch + spec.shard_committee_period;
```
I think we should do a `checked_add` here, in case the `activation_epoch` is set to `FAR_FUTURE_EPOCH`.

paulhauner commented on account_manager/src/validator/exit.rs:
```diff
@@ -0,0 +1,440 @@
+
+/// Get the validator index af given the validator public key by querying the beacon node endpoint.
+/// Returns an error if the beacon endpoint returns an error or given validator is not eligible for an exit.
+async fn get_validator_index<E: EthSpec>(
```
I think we should call this `get_validator_index_for_exit` since we also do verification in here to see if they're eligible.

paulhauner commented on account_manager/src/validator/exit.rs:
```diff
@@ -0,0 +1,440 @@
+) -> Result<Fork, String> {
+    Ok(client
+        // TODO: verify this is the state that we want to query.
+        .get_beacon_states_fork(StateId::Slot(epoch.start_slot(E::slots_per_epoch())))
```
I think we can also use `StateId::Head` here too.

paulhauner commented on account_manager/src/validator/exit.rs:
```diff
@@ -0,0 +1,440 @@
+    let elapsed = current_time.safe_sub(genesis_time)?;
+    let seconds_per_slot = spec.milliseconds_per_slot.safe_div(1000)?;
+
+    let current_slot = Slot::new(elapsed.safe_div(seconds_per_slot)?);
```
This doesn't consider the case where `GENESIS_SLOT > 0`, which whilst unlikely here, will happen for shard chains. The safest thing to do here is to start a `SystemTimeSlotClock` and read it, like in the tests.

pawanjay176 replied:
Thank you. That's much better :)

paulhauner commented on account_manager/src/validator/exit.rs:
```diff
@@ -0,0 +1,440 @@
+
+/// Load the voting keypair by loading the keystore and decrypting the keystore
+///
+/// First attempts to load the password for the validator from the `secrets_dir`, if not
```
```suggestion
/// First attempt to load the password for the validator from the `secrets_dir`, if not
```

paulhauner commented on account_manager/src/validator/exit.rs:
```diff
@@ -0,0 +1,440 @@
+    App::new("exit")
+        .about("Submits a VoluntaryExit to the beacon chain for a given validator(s).")
+        .arg(
+            Arg::with_name(VALIDATOR_FLAG)
```
I would be tempted to just accept a path to an EIP-2335 keystore and always prompt for the password, instead of dealing with the validator dirs. It'll work in a broader set of cases and will be less code for us to maintain.

Now we've done (are doing) away with `lighthouse account validator deposit`, we don't need to follow its lead regarding this flag anymore.

pawanjay176 replied:
Yes I would like this very much. Will make the changes.

pawanjay176 replied:
What do you think about an option for providing a keystore password file. If no file is provided, then we prompt for a password. Could be easier for people trying to exit multiple validators with different passwords.

paulhauner commented on account_manager/src/validator/exit.rs:
```diff
@@ -0,0 +1,440 @@
+                .takes_value(true),
+        )
+        .arg(
+            Arg::with_name(SECRETS_DIR_FLAG)
```
If you drop the `--validator` flag you can also drop this.

paulhauner reviewed (changes requested):
I like the general structure, looking good. I have mostly typo fixes and one major-ish suggestion: to use EIP-2335 keystores directly instead of the "validators dir".

pawanjay176 commented:
@paulhauner Addressed your comments and added a `password-file` option. 
Also added a check for the remote beacon client to be synced.

I'm not sure how to do end to end testing for this. I think it would require us to have some mock beacon client implementation. Is there any other simpler way ?

paulhauner commented on account_manager/src/validator/exit.rs:
```diff
@@ -0,0 +1,347 @@
+        );
+    } else {
+        eprintln!(
+            "Did not publish voluntary exit for validator {}. Please check that you entered the correct passphrase.",
```
```suggestion
            "Did not publish voluntary exit for validator {}. Please check that you entered the correct exit phrase.",
```

paulhauner commented on account_manager/src/validator/exit.rs:
```diff
@@ -0,0 +1,347 @@
+    }
+
+    // Return immediately if beacon node is not synced
+    if !is_syncing(client).await? {
```
```suggestion
    if is_syncing(client).await? {
```

paulhauner commented on account_manager/src/validator/exit.rs:
```diff
@@ -0,0 +1,347 @@
+        .map_err(|e| format!("Failed to get validator details: {:?}", e))?
+        .ok_or_else(|| {
+            format!(
+                "Validator {} is not present in the beacon state. Please ensure that your beacon node is synced",
```
```suggestion
                "Validator {} is not present in the beacon state. Please ensure that your beacon node is synced and the validator has been deposited.",
```

paulhauner reviewed (changes requested):
Looking great, super tidy! I only have a few more things I found after running it again :)

paulhauner reviewed (approved):
Great!

bors r+
